controllers/apps/shardingdefinition_controller.go

```diff
@@ -0,0 +1,294 @@
+	shardingDef.Status.ObservedGeneration = shardingDef.Generation
+	shardingDef.Status.Phase = phase
+	shardingDef.Status.Message = message
+	return r.Client.Status().Patch(rctx.Ctx, shardingDef, patch)
+}
+
+func (r *ShardingDefinitionReconciler) validate(cli client.Client, rctx intctrlutil.RequestCtx, shardingDef *appsv1.ShardingDefinition) error {
+	for _, validator := range []func(context.Context, client.Client, *appsv1.ShardingDefinition) error{
+		r.validateTemplate,
+		r.validateShardsLimit,
+		r.validateProvisionNUpdateStrategy,
+		r.validateLifecycleActions,
+		r.validateSystemAccounts,
+	} {
+		if err := validator(rctx.Ctx, cli, shardingDef); err != nil {
+			return err
+		}
+	}
+	return r.immutableCheck(shardingDef)
+}
+
+func (r *ShardingDefinitionReconciler) validateTemplate(ctx context.Context, cli client.Client,
+	shardingDef *appsv1.ShardingDefinition) error {
+	template := shardingDef.Spec.Template
+
+	if err := component.ValidateCompDefRegexp(template.CompDef); err != nil {
+		return err
+	}
+
+	compDefs, err := listCompDefinitionsWithPattern(ctx, cli, template.CompDef)
+	if err != nil {
+		return err
+	}
+	if len(compDefs) == 0 {
+		return fmt.Errorf("no component definition found for the specified template")
+	}
+
+	r.compDefs = compDefs // carry the component definitions for later use
+
+	return nil
+}
+
+func (r *ShardingDefinitionReconciler) validateShardsLimit(ctx context.Context, cli client.Client,
+	shardingDef *appsv1.ShardingDefinition) error {
+	return nil
+}
+
+func (r *ShardingDefinitionReconciler) validateProvisionNUpdateStrategy(ctx context.Context, cli client.Client,
+	shardingDef *appsv1.ShardingDefinition) error {
+	supported := func(strategy *appsv1.UpdateStrategy) bool {
+		if strategy == nil {
+			return true
+		}
+		return *strategy == appsv1.SerialStrategy || *strategy == appsv1.ParallelStrategy
+	}
+	if !supported(shardingDef.Spec.ProvisionStrategy) {
+		return fmt.Errorf("unsupported provision strategy: %s", *shardingDef.Spec.ProvisionStrategy)
+	}
+	if !supported(shardingDef.Spec.UpdateStrategy) {
+		return fmt.Errorf("unsupported update strategy: %s", *shardingDef.Spec.UpdateStrategy)
+	}
+	return nil
+}
+
+func (r *ShardingDefinitionReconciler) validateLifecycleActions(ctx context.Context, cli client.Client,
+	shardingDef *appsv1.ShardingDefinition) error {
+	return nil
+}
+
+func (r *ShardingDefinitionReconciler) validateSystemAccounts(ctx context.Context, cli client.Client,
+	shardingDef *appsv1.ShardingDefinition) error {
+	if !checkUniqueItemWithValue(shardingDef.Spec.SystemAccounts, "Name", nil) {
+		return fmt.Errorf("duplicate system accounts are specified")
+	}
+
+	for _, account := range shardingDef.Spec.SystemAccounts {
+		if err := r.validateSystemAccountDefined(account.Name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ShardingDefinitionReconciler) validateSystemAccountDefined(name string) error {
+	for _, compDef := range r.compDefs {
```

Comment:
This validation rule checks if the matched systemAccount belongs to a different compDef than the one actually referenced by the sharding？

Reply:
It checks all available component definitions.